src/app/cluster-building-blocks/QuieterReporting.h

```diff
@@ -0,0 +1,194 @@
+
+/**
+ * This class helps track reporting state of an attribute to properly keep track of whether
+ * it needs to be marked as dirty or not for purposes of reporting using Q quality.
+ *
+ * The class can be configured via `SetPolicy` to have some/all of the common reasons
+ * for reporting (e.g. increment only, decrement only, change to/from zero).
+ *
+ * Changes of null to non-null or non-null to null are always considered dirty.
+ *
+ * It is possible to force mark the attribute as dirty (see `ForceMarkAsDirty`) such as
+ * for conditions like "When there is any increase or decrease in the estimated time
+ * remaining that was due to progressing insight of the server's control logic".
+ *
+ * Usage is simple:
+ *
+ * - Call `SetValue()` with the new value and current monotonic system timestamp
+ *   - There is an overload with a `SufficientChangePredicate` which will apply externally
+ *     provided checks between old/new value and time last marked dirty to allow for complex
+ *     rules like marking dirty less than once per second, etc.
+ * - *Maybe* call `ForceMarkAsDirty()` in some choice situations (e.g. know for sure it's
+ *    an important update, like at the edge of an operational state change).
+ * - Call `GetThenResetDirtyState()`. If it returns true, mark the attribute dirty, with the
+ *   method most suitable at the call site (e.g. `MatterReportingAttributeChangeCallback` call
+ *   or similar methods).
+ *
+ * @tparam T - the type of underlying numerical value that will be held by the class.
+ */
+template <typename T>
+class QuieterReportingAttribute
+{
+  public:
+    explicit QuieterReportingAttribute(const chip::app::DataModel::Nullable<T>& initialValue) : mValue(initialValue), mLastDirtyValue(initialValue) {}
+
+    using SufficientChangePredicate = std::function<bool(Timestamp /* previousDirtyTime */, Timestamp /* now */, const Nullable<T> & /* previousDirtyValue */, const Nullable<T> & /* newValue */)>;
+
+    /**
+     * @brief Factory to generate a functor for "attribute was last reported" at least `minimumDurationMillis` ago.
+     *
+     * @param minimumDurationMillis - number of millis needed since last marked as dirty before we mark dirty again.
+     * @return a functor usable for the `changedPredicate` arg of `SetValue()`
+     */
+    static SufficientChangePredicate GetPredicateForSufficientTimeSinceLastDirty(chip::System::Clock::Milliseconds64 minimumDurationMillis)
+    {
+      return [minimumDurationMillis](Timestamp previousDirtyTime, Timestamp now, const Nullable<T> &oldDirtyValue, const Nullable<T> &newValue) -> bool {
+        return (oldDirtyValue != newValue) && ((now - previousDirtyTime) >= minimumDurationMillis);
+      };
+    }
+
+    chip::app::DataModel::Nullable<T> value() const { return mValue; }
+    QuieterReportingPolicyFlags policy() const { return mPolicyFlags; }
+
+    void SetPolicy(QuieterReportingPolicyFlags policyFlags) { mPolicyFlags = policyFlags; }
```

Comment:
Did we consider having:

```cpp
const QuieterReportingPolicyFlags &policy() const { return mPolicyFlags; }
QuieterReportingPolicyFlags &policy() { return mPolicyFlags; }
```

instead and then we can do things like:

```cpp
foo.policy().Set(QuieterReportingPolicyEnum::kMarkDirtyOnDecrement);

// or use clear or even operator=
```

instead of needing

```cpp
foo.SetPolicy()QuieterReportingPolicyFlags(QuieterReportingPolicyEnum::kMarkDirtyOnDecrement));
```

Reply:
Will rework